Support for custom integrations

ergl opened the pull request:
The syntax to add a new custom integration is as follows:

``` yaml
custom_integrations:
    your_integration_name:
        path: "/absolute/path/to/integration/file"
        name: "NameOfIntegrationClass"
```

For each custom integration, the user should supply the path of the
integration file, and the name of the class containing the integration
inside the file (that would be the class inheriting from
AbstractIntegration).

The program will abort if no matching class is found inside the given
file.

Still pending: add tests

ergl committed:
loader.load -> loader.load_test_files, add documentation

ergl committed:
Apply correct formatting on _load_integrations_from_configuration

ergl committed:
Support custom integrations

The syntax to add a new custom integration is as follows:

```
custom_integrations:
    your_integration_name:
	path: "/absolute/path/to/integration/file"
	name: "NameOfIntegrationClass"
```

For each custom integration, the user should supply the path of the
integration file, and the name of the class containing the integration
inside the file (that would be the class inheriting from
AbstractIntegration).

The program will abort if no matching class is found inside the given
file.

angelsanzn commented:
Looks nice! Let's give it a little more thought while we add some tests...

ergl commented:
Tests aside, what do you think of this approach to defining custom integrations?

I know it's a bit different from what we thought at first, but this is the
simplest way I could think of.

The `name` field is there so that we know what class to add to
`stock_integrations.integration_classes`.

A simple custom integration (same behaviour as terminal) would look like:

``` python
from pysellus.interfaces import AbstractIntegration

class MyService(AbstractIntegration):
    def on_next(self, message):
        print('Assert error in {} -> {}'.format(
            message['test_name'],
            message['expect_function']
        ))
```

And the appropiate custom declaration would be:

``` yaml
custom_integrations:
    my_service:
        path: "/.../sample_integration.py"
        name: "MyService"
```

ergl commented:
~~(...) adding 3rd pary integrations to `stock_integrations` is [not] a good idea (...).~~

EDIT 3: Also fixed in latest change

~~Right now we could define a custom integration with
the same name as a stock integration, which would override the original
implementation.~~

~~If a user defines an custom integration and defines it as `slack` in the
`your_integration_name` field of `custom_integrations`, all instances of `slack`
in the `notify` section will reference the custom version.~~

EDIT 2: Fixed in latest commit

angelsanzn commented:
I believe this scheme in the config file is the simplest we can get -- we need at least that info in order to know what to load and how to refer to it later.

angelsanzn commented:
You're right that, semantically, adding instances of custom integrations' classes to the dict in `stock_integrations` is wrong. But I think that's easily solved by having a higher-level dict combining both the stock and the custom integrations.

I don't think the ability to override stock integrations is something valuable to end users: if they want to use a stock integration, they can use them and alias them as they want. And if they don't, they can define their own integrations, potentially importing stock ones from `pysellus.stock_integrations` in case they want to reuse their implementations.

I think it's better to abort if a custom integration is declared with a name equal to any of the stock ones, as you described, but only because that would elimiante all ambiguity as to which integration is to be used.

ergl committed:
Prevent the user from overriding stock integrations

Integration names must be unique. If the program finds two integrations
with the same name, it will exit.

ergl committed:
Improve missing integration error message

ergl commented:
> Right now we could define a custom integration with
> the same name as a stock integration, which would override the original
> implementation.

Fixed this in latest commit.

Also, travis seems to be having problems right now, will push another change in a few hours when it's back again.

ergl committed:
integration_classes -> stock_integration_classes

ergl committed:
Program now aborts if config file is empty

Although not defining any custom integrations is allowed

ergl commented:
Also adding better error handling to `integration_config`, but let me know if that should be better placed in another pull request

ergl committed:
Create higher-level integration dictionary

`integrations.integration_classes` now holds both stock and custom
integrations.

All previous key checks in `integration_config` now test that high level
dict in `integrations`. To prevent circular dependencies,
`integration_config.loaded_integrations` had to be moved to
`integrations`.

Also modified tests accordingly, and improved error message in
`integrations._create`

ergl committed:
Make unpack_integration... private

ergl commented:
> You're right that, semantically, adding instances of custom integrations' classes to the dict in `stock_integrations` is wrong. But I think that's easily solved by having a higher-level dict combining both the stock and the custom integrations.

Changed in latest commit. I think this latest change settles the pr as complete. There shouldn't be any more changes to the code, except adding tests.

ergl committed:
Move config file error handling

ergl committed:
Improve integration_config documentation

ergl committed:
Move _load_defined_integrations definition

ergl committed:
Add initial tests for new integration system

ergl committed:
Additional tests to integrations_spec and integration_config

ergl commented:
Well, I think I'm done for now, I went ahead and added test cases for all the
new things I added, and a few more.

@angelsanz take a look at it and tell me what you think. Before merging I'll
squash a couple of commits so don't merge this for now.

angelsanzn commented on pysellus/integration_config.py:
```diff
@@ -42,19 +39,125 @@ def _load_config_file(path):
+        integration_name = configuration.pop('name', None)
+        integration_path = configuration.pop('path', None)
+
+        if integration_classes is None or integration_path is None:
```
Shouldn't this be `intgration_name` instead of `integration_classes`?

ergl replied:
Oops, you're right, autocomplete bit me there! Should be fixed now.

Commit:
Fix nesting in tests

Commit:
integration_config: refactor to guard clause

I think this is clearer because, before, it wasn't very
clear where the KeyError might come (and it could come
from the call to `_load_custom_integrations_classes` if
that code is ever changed.

ergl committed:
Fix integration_config check

Commit:
integration_config_spec: split a test case in two smaller ones

Commit:
integrations config: extract `_load_configuration_from_contents_of_config_file` method to make testing easier

This allows us to have a unit test instead of creating and deleting
files on each test run. It also gets rid of the noted PermissionError
(although I never had it, I had other errors: nonexistent file or...
existent file - but nevermind).

Commit:
integration_config_spec: amend the purpose of some integration tests

The two tests verifying the behaviour when the config file
doesn't exist were passing because the bogus paths obviously
made the program raise FileNotFoundError. But we wanted to check
that, directories that actually exist, but do not contain a config
file, that exception is raised. This patch changes those tests
to do just that.

This patch also refactors integration_config, extracting a method
from `_load_config_file`.

Commit:
integration_config_spec: refactor a test to use a loop

Commit:
integration_config_spec: split a test into two simpler ones

In the original test, two exercises were being made. This
separates each exercise into its own test. This also extracts
a method in integration_config to encapsulate the call to
`inspect`, hence allowing us to stub that method instead of
the direct call to the `inspect` module.

Commit:
integration_config_spec: add a missing 'teardown' step in a test

Even though the tests weren't failing, I believe it's necessary
to undo the 'manual stub' in this test.

Commit:
Refactor integration_config

Turn a check for falsiness into a check for the concrete exception being thrown
by default, instead of hardcoding a default return value an then checking for
its presence.

Commit:
integrations: narrow the scope of a try-except clause

Commit:
Comply with flake8

ergl committed:
_load_conf..._contents... -> _load_configuration_from_config_file
